backends/tfhe-cuda-backend/cuda/src/crypto/keyswitch.cuh

```diff
@@ -146,12 +147,188 @@ void execute_keyswitch_async(cudaStream_t *streams, uint32_t *gpu_indexes,
+}
+
+// public functional packing keyswitch
+//
+// blockIdx.y - input
+// chunk_count = glwe_size * polynomial_size / threads.
+template <typename Torus>
+__global__ void
+fp_keyswitch_lwe_list_to_glwe(Torus *glwe_array_out, Torus *lwe_array_in,
+                              Torus *fp_ksk, uint32_t lwe_dimension_in,
+                              uint32_t glwe_dimension, uint32_t polynomial_size,
+                              uint32_t base_log, uint32_t level_count,
+                              Torus *d_mem) {
+  const int glwe_accumulator_size = (glwe_dimension + 1) * polynomial_size;
+  const int lwe_size = (lwe_dimension_in + 1);
+
+  const int input_id = blockIdx.y;
+  const int degree = input_id;
+
+  // Select an input
+  auto lwe_in = lwe_array_in + input_id * lwe_size;
+  auto ks_glwe_out = d_mem + input_id * glwe_accumulator_size;
+  auto glwe_out = glwe_array_out + input_id * glwe_accumulator_size;
+  // KS LWE to GLWE
+  keyswitch_lwe_ciphertext_into_glwe_ciphertext(
+      ks_glwe_out, lwe_in, fp_ksk, lwe_dimension_in, glwe_dimension,
+      polynomial_size, base_log, level_count);
+
+  // P * x ^degree
+  polynomial_accumulate_monic_monomial_mul_batch(
+      glwe_out, ks_glwe_out, degree, polynomial_size, glwe_dimension + 1, true);
+}
+
+/// To-do: Rewrite this kernel for efficiency
+template <typename Torus>
+__global__ void accumulate_glwes(Torus *glwe_out, Torus *glwe_array_in,
+                                 uint32_t glwe_dimension,
+                                 uint32_t polynomial_size, uint32_t num_lwes) {
+  const int tid = threadIdx.x + blockIdx.x * blockDim.x;
+
+  extern __shared__ int8_t sharedmem[];
+  const int shmem_index = threadIdx.x;
+
+  Torus *glwe_acc = (Torus *)sharedmem;
+  glwe_acc[shmem_index] = glwe_array_in[tid];
+
+  // Accumulate
+  for (int i = 1; i < num_lwes; i++) {
+    auto glwe_in = glwe_array_in + i * (glwe_dimension + 1) * polynomial_size;
+    glwe_acc[shmem_index] += glwe_in[tid];
+  }
+
+  glwe_out[tid] = glwe_acc[shmem_index];
+}
+
+template <typename Torus>
+__host__ void host_fp_keyswitch_lwe_list_to_glwe(
+    cudaStream_t stream, uint32_t gpu_index, Torus *glwe_out,
+    Torus *lwe_array_in, Torus *fp_ksk_array, uint32_t lwe_dimension_in,
+    uint32_t glwe_dimension, uint32_t polynomial_size, uint32_t base_log,
+    uint32_t level_count, uint32_t num_lwes) {
+  cudaSetDevice(gpu_index);
+  int glwe_accumulator_size = (glwe_dimension + 1) * polynomial_size;
+
+  int num_blocks = 0, num_threads = 0;
+  getNumBlocksAndThreads(glwe_accumulator_size, 512, num_blocks, num_threads);
+
+  auto shared_mem = sizeof(Torus) * num_threads;
+  dim3 grid(num_blocks, num_lwes);
+  dim3 threads(num_threads);
+
+  auto d_mem = (Torus *)cuda_malloc_async(
+      num_lwes * glwe_accumulator_size * sizeof(Torus), stream, gpu_index);
+  auto d_tmp_glwe_array_out = (Torus *)cuda_malloc_async(
+      num_lwes * glwe_accumulator_size * sizeof(Torus), stream, gpu_index);
```

Comment:
Maybe these could be allocated via the scratch?

Reply:
Definitely. This is part of the polishment I've mentioned. Will do it.

Reply:
Done.